### DISCUSSION TITLE
Is there a recommended way to pass information to a layout when a new window is added?

### DISCUSSION THREAD

**Discussion by aravinda0** [qtile/qtile#3478] [Q&A] [answered]
From what I gather, `Layout.add(window)` is invoked when a new window is created, so the `Layout` class can capture it and store internally if needed.

I'm trying to code up a tmux-esque layout where one can dynamically specify via bindings if we want to open a window as a 'vsplit' or a 'hsplit'. 

A couple approaches came to mind but wanted to get feedback. Assume I have a helper called `lazy_vsplit(shell_cmd)` that I can bind to some hotkey. 

#### 1. Global state approach

`lazy_vsplit()` will:

- Set `Layout.window_add_mode = vsplit`
- Trigger the window creation with passed in cmd - `lazy.spawn(shell_cmd)`
- Later, the `Layout.add()` override will capture new window and associate it with the current `window_add_mode` as the requested way to lay it out

Usual awkwardness around global state. Could work unless this flow involes some race condition wrt how window creation happens.

#### 2. Environment Variables?

`lazy_vsplit()` will basically decorate `lazy.spawn(shell_cmd)`.

So `shell_cmd = f"QTILE_MY_LAYOUT_ADD_MODE=vsplit {shell_cmd}"`

Would this allow me to somehow access this env  var from the window reference during `Layout.add(window)`?

#### 3. Some other way?

Is there some more qtiley way of doing this?

**Comment by elParaguayo** [comment, marked as answer]
I'd avoid option 2!

Option 1 makes sense. If your layout has different modes then all your function needs to do is set the mode first and add the window. Provided the mode is set (and your layout code works) then the layout should position the next window correctly.

Have you seen the Plasma layout (https://github.com/numirias/qtile-plasma) it might be similar to what you're looking at.

I'm also working on something similar but I've been focussing on other things recently.

**Reply by aravinda0** [reply]
Thanks for your input, and for that link to qtile-plasma - I didn't know about it. That project does look similar to what I'm trying to build - though I'm trying to take things one further by also allowing tabs and even tabs within split panes.

It looks like qtile-plasma also uses option 1, so I'll go with that too.
